tests/js/server/aql/aql-optimizer-geoindex.js

```diff
@@ -1029,7 +1029,64 @@ function optimizerRuleTestSuite() {
       var result = AQL_EXPLAIN(query.string, query.bindVars);
       hasIndexNode(result, query);
       hasNoFilterNode(result, query);
-    }
+    },
+
+    // Regression test (https://arangodb.atlassian.net/browse/AR-113):
+    // Previously, the filter was moved into the index, but one access of `dist` in `dist < dist` was left, while the
+    // variable obviously can't be available yet at the index node.
+    // When failing, throws an error like
+    //   missing variable #1 (dist) for node #8 (IndexNode) while planning registers
+    testSelfReference1: function () {
+      const query = {
+        string: `
+          FOR doc IN @@col
+            LET dist = GEO_DISTANCE(doc.geo, [16, 53])
+            FILTER dist < dist
+            RETURN {doc, dist}
+        `,
+        bindVars: {'@col': colName},
+      };
+
+      const result = AQL_EXPLAIN(query.string, query.bindVars);
+      hasNoIndexNode(result, query);
+      hasFilterNode(result, query);
+    },
+
+    // See testSelfReference1
+    testSelfReference2: function () {
+      const query = {
+        string: `
+          FOR doc IN @@col
+            LET dist = GEO_DISTANCE(doc.geo, [16, 53])
+            LET dist2 = GEO_DISTANCE(doc.geo, [3, 7])
+            FILTER dist < dist2
```

Comment:
So my understanding here is that we can't use the index to do the calculation here because it's two-sided. Essentially, we don't have some sort of fixed query point to feed to the index. Right?

Reply:
I guess what I'm asking is, is the only reason we got into trouble before these changes (i.e. the only time we get invalid vars) is when they are dependent on the document being fetched by the index? If there's another way to generate such an invalid var without it depending on the document being served by index (if it's used), then I think we need to write a test for it.

Reply:
You're right, other variables that are defined before the filter but after the enumerate collection node are also affected. I'll write another test that does this.

Reply:
Yeah, I mean I definitely understand how these existing cases would trip the same rule, just I can see where somehow we might detect them separately in the future and want to make sure we still catch the cases for variables that have nothing to do with the index access itself.

Reply:
Writing such a test is sadly currently not possible, until [BTS-256](https://arangodb.atlassian.net/browse/BTS-256) is also fixed.